pingroups-proposal.md

```diff
@@ -0,0 +1,122 @@
+# Digital Pin Groups (Proposal)
+
+Provides support for the situation where you want to set or get the values of 
+an arbitrary set of digital IO pins that may not necessarily align to a port and 
+do all of this in one operation.
+
+Examples of this behaviour would include:
+
+* Analog Multiplexer / Demultimplexer, where you need to set the bit value of
+three pins in order to determine which analog line is being used on the multiplexer.
+* Keypads where the value of the key presses are expressed using a combination of
+states across a set of digital lines (eg: [https://www.sparkfun.com/products/8653](https://www.sparkfun.com/products/8653) )
+
+When you want to issue an equivalent of a digitalWrite to a group of pins,
+you'll then issue a 7-bit byte that provides the states of the pins collectively. 
+This will save several calls to digital write and allow them to be done in one group.
+
+Reads will work the same way but return a byte with the states of all of the pins.
+This is particularly important in a scenario like a keypad where independent
+async reads can make it extremely challenging to get the state of the keypress properly.
+
+## Requirements
+
+* Currently unimplemented (PoC to come shortly)
+* An array of pin groups (suggest 8 groups so it can be identifed with 3 bits 
+each with 7 pins defined in the group)
+* Modifications to firmata to accept the new protocol.
+
+## Protocol
+
+### Digital Pin Group commands
+
+In order to save space in the protocol, the Digital Pin Group command comprises
+both protocol commands as well as the id address space for the groups as below:
+
+LSB
+0 - 2:  3 bits to determine which Pin Group command is being issued
+3    :  Reserved for future use / address space increases
+4 - 6:  3 bits for Pin Group ID address space - providing up to 8 distinct digital pin groups
+
+Command values are provided below
+```
+CONFIG              (0x00)
+CLEAR               (0x01)
+PIN_STATE_SET       (0x02)
+PIN_STATE_REQUEST   (0x03)
+PIN_STATE_REPLY     (0x04)
+future reserved     (0x05 - 0x07)
+```
+
+### Configuration
+
+Specifies which pins should be grouped together and in which order. A maximum
+of 7 pins can be grouped together in one pin group. When specified in the config
+message, the pins will be provided in little endian style so the first pin will
+then be configured to mapped to the Least Significant Bit in subsequent write
+and read messages.
+
+```
+0:  START_SYSEX         (0xF0)
+1:  pin group command   (0x60)
+2:  pin group id (0 - 7) << 4 | CONFIG
+3:  first pin in pin group (0 - 127)
+4:  second pin in pin group (0 - 127)
+... up to maximum of 7
+N:  END_SYSEX           (0xF7)
+```
+
+### Clear
+
+Clears the values of a given pin group
```

Comment:
Got it, so the user would send the clear message and then send another config message with the same ID but perhaps with more or less pins this time (and different pins even).

Could also do this automatically internally if someone calls config on an ID a second time, call clear internally and then re-config.